Use CacheControl module for better caching

b-m-f opened the pull request:
## DONE
Implements 1 part of https://github.com/ubuntudesign/base-squad/issues/119
Fixes #5, #6 

## QA
Here is a test script to make sure that the code works.
This needs a Redis server running at the standard config locally.
```python
import time
import redis
from cachecontrol.caches.file_cache import FileCache
from cachecontrol.caches.redis_cache import RedisCache
from canonicalwebteam.http import CachedSession
from datetime import datetime


def create_redis_connection_pool(
    redis_port="6379", redis_host="localhost", redis_db=0
):
    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)


pool = create_redis_connection_pool()
sess = CachedSession(redis_cache_pool=pool, fallback_cache_duration=10)
print(sess)
for i in range(9):
    print("start")
    print(datetime.now())
    response = sess.get("https://ehlers.berlin")

    print("is response from cache")
    print(response.from_cache)
    print(datetime.now())
    print(response.headers)
    time.sleep(10)
```

b-m-f committed:
adds CacheControl library for cached sessions, with a default heuristic
when Headers are missing.

Greggless commented:
# -1

tbille commented on canonicalwebteam/http/__init__.py:
```diff
@@ -5,28 +5,29 @@
+from requests import adapters, session, exceptions, Session
+from cachecontrol import CacheControl
+from cachecontrol.caches.redis_cache import RedisCache
+from cachecontrol.heuristics import ExpiresAfter
```
re-order imports

Greggless replied:
I should

b-m-f replied:
Is done

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,16 @@
+
+
+if __name__ == "__main__":
+    unittest.main()
```
If you want to have this you should `import unittest`.

I try to run usually tests with `python -m unittest discover tests` which doens't need this block.
So you can remove those two lines :smile:

nottrobin replied:
I would suggest leaving those two lines. Providing people with the option to run your python tests directly is simply good manners.

But yes, if you want to use `unittest`, you need to `import` it.

tbille commented on setup.py:
```diff
@@ -4,18 +4,20 @@
-    author_email='webteam@canonical.com',
-    url='https://github.com/canonicalwebteam/http',
+    name="canonicalwebteam.http",
+    version="0.1.6",
```
in this PR you will need to upgrade the version and the [CHANGES.txt](https://github.com/canonical-webteam/http/blob/master/CHANGES.txt)

tbille commented on setup.py:
```diff
@@ -4,18 +4,20 @@
+    version="0.1.6",
+    author="Canonical webteam",
+    author_email="webteam@canonical.com",
+    url="https://github.com/canonicalwebteam/http",
```
not from you but link is wrong https://github.com/canonical-webteam/http (- missing)

tbille commented:
There is a yolo file `test` :smile:

tbille commented on README.rst:
```diff
@@ -16,3 +16,7 @@ This module is in PyPi as :code:`canonicalwebteam.http`. You should be able to i
+
+Tests
+-----
+Tests are located in the `tests` directory and can be run with `python -m unittest discover tests/`.
```
A line before

tbille commented on canonicalwebteam/http/__init__.py:
```diff
@@ -5,28 +5,29 @@
+from cachecontrol.caches.redis_cache import RedisCache
+from cachecontrol.heuristics import ExpiresAfter
+import redis
+from requests import adapters, session, exceptions, Session
```
The way python orders works is:
```python
import a
import b
from a import d
from b import s
```

tbille replied:
like this:
```python
try:
    from urllib.parse import urlparse
except ImportError:
    from urlparse import urlparse

from requests import Session, adapters, exceptions, session

import redis
from cachecontrol import CacheControl
from cachecontrol.caches.redis_cache import RedisCache
from cachecontrol.heuristics import ExpiresAfter
```

nottrobin replied:
@tbille I agree that that looks neat, but do you have a source for it being any sort of standard?

PEP8 [mentions](https://www.python.org/dev/peps/pep-0008/#imports):

> Imports should be grouped in the following order:
> 
> Standard library imports.
> Related third party imports.
> Local application/library specific imports.
> You should put a blank line between each group of imports.

So you should try to follow those guidelines.

Putting imports in alphabetical order, and putting `from` lines after `import` lines seems to be [a common practice](https://stackoverflow.com/a/20763446/613540) (and one that we could mention in our practices, if we like), but I'm curious if it is written down anywhere as an actual standard.

tbille commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +101,31 @@ class UncachedSession(BaseSession, requests.Session):
 
 
-class CachedSession(BaseSession, requests_cache.CachedSession):
+def create_cached_session(
```
as discussed we should be consistant, or we use classes, or we use factories for both of the sessions. I think that in this case I'd rather use classes.

nottrobin replied:
I'd like to be included in that discussion. I advised @b-m-f to use a function (a "factory function" that returns a `BaseSession` object) here, as there doesn't seem to be a neat way to subclass both our `BaseSession` (extension of `requests.Session`) and `CacheControl` together. If there is a neat way to create that class that doesn't lead to an object with a rather confusing type, then I'm happy to go with that.

But if the best way is, as I suspect, to simply pass our base session to the `CacheControl` class on initiation, then I think this should remain a function as that's the simplest construct. Creating a class for the sake of it is unnecessary complexity.

b-m-f replied:
I have updated this. Using the adapter seems to work neatly. I have outlined the problem of overwriting the TimeoutAdapter below. 
If needed I could dig deeper to see if chaining is possible somehow.

tbille commented:
@b-m-f I am not sure if we want this in this PR but: we should also provide a way for the user to allow him not to use redis (in dev mode for ex) but keep the file cache system that we had.

@nottrobin any opinion on this?

tbille commented:
You can check this file https://github.com/canonical-websites/snapcraft.io/blob/20efef5cb66e2306d17330e9ddf84dc8c25b1de9/webapp/api/requests.py

It evolved a bit since we created this repo

b-m-f commented:
> @b-m-f I am not sure if we want this in this PR but: we should also provide a way for the user to allow him not to use redis (in dev mode for ex) but keep the file cache system that we had.
> 
> @nottrobin any opinion on this?

We could add two functions/classes for this. Standard cache and redis. Or pass an option to the function, that the dev could overwrite in the implementation?

nottrobin commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +101,31 @@ class UncachedSession(BaseSession, requests.Session):
-
-    Responses for an identical request will be naively returned from the
-    cache if the cached copy if less than "expire_after" seconds old.
+    A factory function for making HTTP requests with cached responses.
```
How about:

```suggestion
    A factory function to instantiate a CacheControl session with
    our default settings from BaseSession.
```

b-m-f replied:
Updated it

nottrobin commented:
> @tbille: I am not sure if we want this in this PR but: we should also provide a way for the user to allow him not to use redis (in dev mode for ex) but keep the file cache system that we had.

I agree. I actually think the default should remain as a file-based cache, with the option to pass a redis cache object.

> @b-m-f: We could add two functions/classes for this. Standard cache and redis. Or pass an option to the function, that the dev could overwrite in the implementation?

I think the function you've created should remain, with simply an optional parameter for the RedisCache. If it's missing, then use the file-based cache.

I also think that the logic of setting up the connection (turning `redis_port`, `redis_host`, `redis_db` into a `RedisCache`) should be taken out of that function. Keeping this logic outside the function makes the function more single-purpose and improves the options for dependency injection.

So something like:

``` python
def create_cached_session(
    redis_cache=None,
    fallback_cache_duration=5,
):
```

We could then initially leave it up to the client to do:

``` python
redis_cache = RedisCache(
    connection_pool=redis.ConnectionPool(
        host=redis_host, port=redis_port, db=redis_db
    )
)
session = create_cached_session(redis_cache=redis_cache)
```

This doesn't feel too verbose to me, but if you feel that's too noisy you could create another helper function for creating that `redis_cache`.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,12 @@
+        adapter = session.get_adapter("http://")
+        heuristic = adapter.heuristic
+        delta = heuristic.delta
+        self.assertEqual(delta, timedelta(seconds=10))
```
@b-m-f and @tbille - did you see my comment about putting the `main` lines back? I think including them is better practice. Not everyone coming to our project might know of `unittest`, they might not have the `unittest` binary installed, and they might not know how to do `python -m unittest`. I think it is inherently more thorough and discoverable to allow our tests to be run directly with `python3 test.py`.

tbille commented:
> What were you referring to here? Is there something in that file we should be porting over?

@nottrobin I am just mentioning this file since I think there were some evolutions that could be interesting for this PR. 

> I also think that the logic of setting up the connection (turning redis_port, redis_host, redis_db into a RedisCache) should be taken out of that function. Keeping this logic outside the function makes the function more single-purpose and improves the options for dependency injection.

Should we create a new instance of redis for every CacheSession instance that we create? In snapcraft for example we have 2 instances: blog and store. Should we not instanciate it at the load of the app and then use the same instance for every sessions?

nottrobin commented:
> Should we create a new instance of redis for every CacheSession instance that we create? In snapcraft for example we have 2 instances: blog and store. Should we not instanciate it at the load of the app and then use the same instance for every sessions?

Yes, you probably want to do that. You might also want to share and use the same `cache_session` itself. Either way, the format I suggest enables you to do what makes the most sense - e.g. you could create `redis_cache = RedisCache(..)` in a global space of some sort (settings.py?), and then pass it through to individual `create_cached_session(..)` calls as needed.

nottrobin commented on setup.py:
```diff
@@ -4,18 +4,21 @@
-    author_email='webteam@canonical.com',
-    url='https://github.com/canonicalwebteam/http',
+    name="canonicalwebteam.http",
+    version="0.1.7",
```
Didn't we talk about making this `1.0.0`? I think this is a fairly fundamental change to how the module works, and probably warrants a new major version. Or if you don't think we're ready for v1 yet, we could make it `0.2.0`, but it certainly feels like more than a point release.

b-m-f replied:
Youre right. This slipped my mind. Updated

b-m-f commented:
> > @tbille: Should we create a new instance of redis for every CacheSession instance that we create? In snapcraft for example we have 2 instances: blog and store. Should we not instanciate it at the load of the app and then use the same instance for every sessions?
> 
> Yes, you probably want to do that. You might also want to share and use the same `cache_session` itself. Either way, the format I suggest enables you to do what makes the most sense - e.g. you could create `redis_cache = RedisCache(..)` in a global space of some sort (settings.py?), and then pass it through to individual `create_cached_session(..)` calls as needed.

This is the way I have implemented it now. I also reverted to a Class based approach for consistency in the file. 
The fallback is now a File cache and the file to be used can be specified via a parametern.
One thing to note is that the TimoutAdapter will be overwritten.

nottrobin commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +103,42 @@ class UncachedSession(BaseSession, requests.Session):
-        *args,
+        redis_cache_pool=None,
+        fallback_cache_duration=5,
+        file_cache_name=".webcache" * args,
```
I think you're missing a comma, and probably have an extra space.

tbille commented on README.rst:
```diff
@@ -5,7 +5,7 @@ A Python library for making HTTP requests with sensible defaults
 .. image:: https://travis-ci.org/canonicalwebteam/http.svg?branch=master
    :alt: build status
-   :target: https://travis-ci.org/canonicalweeam/http
+   :target: https://travis-ci.org/canonicalwebteam/http
```
`canonical-webteam`

tbille commented on README.rst:
```diff
@@ -5,7 +5,7 @@ A Python library for making HTTP requests with sensible defaults
 
 .. image:: https://travis-ci.org/canonicalwebteam/http.svg?branch=master
```
`canonical-webteam`

nottrobin commented:
> This is the way I have implemented it now. I also reverted to a Class based approach for consistency in the file.

:+1: on all that. I like the new way. It's neat. Better than the factory function (thanks @tbille).

> The fallback is now a File cache and the file to be used can be specified via a parametern.

:+1:

> One thing to note is that the TimoutAdapter will be overwritten.

I'm afraid we'll need to find a way around this. We still need to be able to successfully set a timeout for our connections. Without that, our apps could hang indefinitely while waiting for API response, which is potentially fatal.

tbille commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +103,43 @@ class UncachedSession(BaseSession, requests.Session):
+    redis_cache_pool -- Port of your Redis instance
+    file_cache_name -- Name for the file to store cache on
+    fallback_cache_duration -- Duration in seconds for fallback caching
+        retention when no CacheControl headers are set
```
We usually use [reStructuredText Docstring Format](https://www.python.org/dev/peps/pep-0287/) for documentation.

I don't mind we change it in the future, but in the meantime we should try to keep to have some consistancy between projects

nottrobin replied:
I suspect all you mean is that you use that format in snapcraft.io. I have never used that format, and I don't remember seeing it anywhere else, so I don't really think it's entirely fair for you to suggest it's a soft standard in our team. I've never used a specific format for denoting parameters.

However, if @b-m-f wants to use it, I have no objection either.

b-m-f replied:
I do actually prefer the `rst` style over the google version as well.  Got [this stackoverflow](https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format) answer from toto. Gives a good overview of the different styles.

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,46 @@
+def create_redis_connection_pool(
+    redis_port="6379", redis_host="localhost", redis_db=0
+):
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
```
What if I don't run redis locally?

Can you check if you can mock redis to run the tests?

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,46 @@
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
+
+
+class TestCachedSession(TestCase):
```
```suggestion
class TestCachedSession(unittest.TestCase):
```

b-m-f replied:
Its imported directly as well. Is that a pattern that is accepted? Or should I do
```python
from unittest import unittest, TestCase
```
or just
```
import unittest
```

instead?

b-m-f replied:
Ok, the first one does not work.

Have updated to `import unittest`

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,46 @@
+from cachecontrol.caches.redis_cache import RedisCache
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+from unittest import TestCase
```
remove this import and  check suggestion further down.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,46 @@
+    def test_default_heuristic(self):
+        session = CachedSession()
+        adapter = session.get_adapter("http://")
+        heuristic = adapter.heuristic
```
Rather than deep-inspecting heuristics, it would be much better to actually test the external behaviour of the module.

In my initial experiments with how to do this, I tried to use [responses](https://github.com/getsentry/responses) (which we've used before to mock HTTP responses), and [requests-mock](https://requests-mock.readthedocs.io/en/latest/). Neither of these worked - they seem to override the caching adapter.

But then I tried [httpretty](https://httpretty.readthedocs.io/en/latest/), which works a charm. E.g.:

``` python
httpretty.register_uri(
    httpretty.GET,
    "https://httpbin.org/ip",
    body='{"origin": "127.0.0.1"}',
    adding_headers={"Cache-Control": "max-age=30"},
)
```

Could you try using this to test the library from the outside?

nottrobin replied:
I like the idea (from httpretty README) of mocking `https://httpbin.org` endpoints, so you can write tests in such a way that they would still work even without the mocking layer. That seems neat to me.

b-m-f replied:
Used it, works great. Throws some errors during test execution though. Specifically:
```
ResourceWarning: unclosed file <_io.BufferedRandom name=9>
  self.fd = FakeSockFile()
```

I have opened an issue at https://github.com/gabrielfalcao/HTTPretty/issues/368

nottrobin commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +103,43 @@ class UncachedSession(BaseSession, requests.Session):
     ):
+        super(BaseSession, self).__init__(*args, **kwargs)
+
+        heuristic = ExpiresAfter(seconds=fallback_cache_duration)
```
Could you write tests to make sure this expiry can be overridden by headers?

Basically, if a response contains `Cache-Control: no-cache` or `Cache-Control: max-age=31622400`, then the instruction from the header should be used instead of our fallback. I suspect that this heuristic currently overrides the header instead.

b-m-f replied:
I have added support for this and added tests.

b-m-f committed:
Adds a custom Heuristic to respect cache-control headers if present

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+                print(resp.headers[h])
+
+        self.assertIn(
+            "Automatically cached for 0:00:10", resp.headers["warning"]
```
Here you're definitely testing internals, and not testing externals.

The "warning" header and its contents are completely custom items that happen to be added by CacheControl. They are of no functional use to the client, they don't prove that the response will actually be cached as desired, and testing them means that if ever anyone swaps out CacheControl for a different module, or possibly simply if CacheControl is upgraded, the test will fail, even though the module may well still be operating exactly correctly.

What I envisaged here was to make multiple `session.get` calls to the same URL, and check the response was returned from the cache or not, as expected. You could use the `from_cache` property, but even that is a little bit testing internals. Even better would be to update the content returned by the endpoint. Maybe we could write a custom `body=` callback function that returns the epoch time? We could then easily prove that the request was a cached version, or not.

nottrobin replied:
Oooh, http://now.httpbin.org/ is a good endpoint for these tests, which we could then mock similarly to how it works live =)

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertIn(
+            "Automatically cached for 0:00:05", resp.headers["warning"]
```
Same here - this is testing internals and not testing externals.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("max-age=30", resp.headers["cache-control"])
```
Again, checking that the header exists is a no-brainer - it proves nothing. What we need to check is that the response is actually cached. Similarly to as described above.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("no-cache", resp.headers["cache-control"])
```
Same point here - the existence of the `cache-control: no-cache` header in no way proves that the caching library won't cache the response.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+        session = CachedSession()
+        adapter = session.get_adapter("http://")
+        cache = adapter.cache
+        self.assertIsInstance(cache, FileCache)
```
A `FileCache` being attached as an adapter to the session proves doesn't prove it will work, and the lack of that adapter doesn't prove it won't work.

What we actually need to test:

1. That the module successfully caches responses in the expected file
2. That the module retrieves the content from that file when returning cached responses (for this, you'd probably have to manipulate the content of the file somehow)

You could either use a mock filesystem, or simply the real filesystem, I don't really mind which. Or if you want to skip this test for now, we can do that, and implement it after merging this PR.

b-m-f replied:
I have updated this test. The problem that I see here is that I again have knowledge in my Test about how the library encodes the data. 
[`CacheControl` is using msgpack](https://github.com/ionrock/cachecontrol/blob/master/cachecontrol/serialize.py#L72). These are again internals. 

I have tried to get a bit of decoupling by using `str()` on both the epoch (comparing the whole response would need proper deserialization) and the cached file before comparing them.

b-m-f replied:
Also, I am not sure how to manipulate the file smartly without the knowledge of the library

nottrobin replied:
It occurred to me - you could avoid having any special knowledge about the data format by doing something like the following:

- Create `CachedSession` using `dir1`, make a request with a long cache time
- Check `dir1` exists
- Create `sess2 = CachedSession` using `dir2`, make the same request, check it gets fresh data
- Delete `dir2`, use `sess2` to make the same request - check you get fresh data
- Create a new `CachedSession` using `dir1` again, check you get the cached original request

This doesn't require inspecting or manipulating the files themselves at all.

I imagine you could use a similar technique for Redis, by creating 2 different mock Redis connections.

b-m-f replied:
Oh yeah, this should work nicely. Thanks! I wont be able to manipulate data, and see that I am getting my modified response, but I think all necessary functionality will still be covered.

b-m-f replied:
I have updated it for both caching backends. 
The redis version needed an update to the Mock Wrapper to differentiate between multiple redis instances.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,105 @@
+        session = CachedSession(redis_cache_pool=pool)
+        adapter = session.get_adapter("http://")
+        cache = adapter.cache
+        self.assertIsInstance(cache, RedisCache)
```
Same here - the `RedisCache` adapter doesn't prove it will actually use Redis, and the lack of one doesn't prove it won't.

Again, we actually need to test:

1. That the module successfully caches responses in the provided Redis
2. That the module retrieves the content from the provided Redis when returning cached responses (so we'd probably have to manually manipulate the content in Redis to test this)

I'm sure Python Redis mocks exist, but this is probably fairly involved, so I'd accept this PR without this test for now if you like.

b-m-f replied:
Implemented it with a **redis-mock** library. One thing I had to add is a wrapper that ensures that the mocked redis object is a singleton for all consumers. This allows the manipulation of the cache entries from outside the library. 

I still have the same problem here though. The library could encode the cached response in any way. Ive tried to directly manipulate the data at a byte level here. 
This should be fairly flexible.

nottrobin reviewed (changes requested):
Ideally tests should confirm that the external, expected behaviour of the module is functioning correctly, but avoid testing anything about the internal implementation ([a good piece on this here](http://blog.ploeh.dk/2015/09/22/unit-testing-internals/)).

They need to make sure the external interfaces and functions work correctly, obviously because that's what the client who uses the module really cares about. If you just test internal logic then there's no guarantee you haven't missed something, and the internal configuration you're testing doesn't actually create the external behaviour you wanted.

And you shouldn't test the internals if you can possibly avoid it because it makes the code very inflexible. To make the module as maintainable as possible, developers should have the option to swap out libraries or change internal logic, as long as the external behaviour of the module remains functional. If you test the internal logic, you force developers to rewrite tests every time they change anything, making them do twice as much work. Which often means that they become unwilling to work on the code, so the module falls into disrepair.

I'm afraid I feel that all your tests here are testing internals and failing to effectively test externals. I've added comments inline.

b-m-f committed:
Improves test for FileCache

b-m-f committed:
further improves testsuite with redis mocking

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,215 @@
+            file_cache_directory=file_cache_directory,
+        )
+
+        # with a 2s retention, and a 1.1s time between requests, 2 of the
```
We do want our tests to run as quickly as possible. Have you tried with even smaller values? If it allows you, it would be good to test with fractions of a second, since I imagine the execution time between calls will only be nanoseconds. Would it let you, for instance, make the `fallback_cache_duration=0.2`? And then sleep for `0.1`?

b-m-f replied:
I have tried it, but unfortunately the cache did not seem to work. 

[The Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#Expiration) say that it is given in seconds, so sub-second values might not work correctly. My best guess here.

b-m-f replied:
I have noticed that the tests become unreliable with such low settings. Without deep investigation of this I would assume it is the unpredictability of IO events that take place in the background.

To get consistent results I have upped the sleeps a bit again. The test will now take 2 seconds. All in all the test-suite takes ~10 seconds. I think it is still acceptable for a module that should not be touched daily.

b-m-f replied:
It seems that the tests still perform inconsistently.
I can not figure it out at the moment, a second pair of eyes on this would be good. Maybe a pairing session.
The false negatives could be really annoying further down the line.

b-m-f replied:
Fixed by using [freezegun](https://github.com/spulec/freezegun)

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,215 @@
+        # the first requests gets send at t=0
+        response = []
+
+        for i in range(3):
```
When we're talking about only 3 iterations, I personally find this unnecessarily complex to read. It barely saves any lines over the procedural:

``` python
response_1 = session.get("https://now.httpbin.org")
time.sleep(1.1)
response_2 = session.get("https://now.httpbin.org")
time.sleep(1.1)
response_3 = session.get(""https://now.httpbin.org)
```

(5 lines vs 4) but I find the procedural version significantly more straightforward to read and understand.

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,215 @@
+
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
```
Since this is only relevant to the `test_file_cache` function, why not do the teardown step in that function itself?

b-m-f replied:
This is the default caching mechanism in the CachedSession. So every function in the tests uses the file cache.
I could switch this to an in-memory cache (maybe a configuration parameter to the CachedSession), but I wanted to replicate the default situation that a running app will have.

If we never use the in-memory based approach via a DictCache outside of testing though, maybe it is overloading the module a bit if I add this?

b-m-f committed:
Updates the tests to better wrap the module logic

nottrobin commented:
> @jpmartinspt: I feel like we might only need a simple factory for cached sessions in here. Basically a simple function that decorates BaseSession with CacheControl and allows us to pass any necessary params to the CacheControl decorator.

I don't quite follow what you mean here - could you illustrate this with an example? Pseudocode would do fine.

> I would not do anything storage specific in here(at least for now), except maybe specify the default cache type we want to use(by setting a default argument in the hypothetical factory function). I believe the default behavior of CacheControl should be sufficient to get us started!

Again, could you illustrate what you mean here? The module at least needs to continue to work how it currently does - i.e. with file storage - and support Redis (which is the point of this task).

> We can implement different caching strategies per project by adding heuristics on the projects themselves and extract them in here if we ever need it.

The point of this module is to create a set of standard defaults for all our projects to follow. I'm hoping that individual projects will very rarely need to diverge from these defaults. They should avoid doing so if possible.

It is also absolutely essential that this module continues to cache all requests for a short period of time, as it currently does. Almost all of our API endpoints contain no `cache-control` headers at the moment (we're hoping they can start to be added once we have the ability to understand them). If we don't add any heuristics by default, then all API calls will not be cached at all. This *will* lead to tied up process and ultimately unresponsive pods and websites.

> after having looked very briefly at the test coverage of the cachecontrol package, I'd say we probably don't need to be adding a lot of these tests since we are probably just replicating what they have tested already?

I do agree we shouldn't be wholesale replicating existing tests and so as long as CacheControl is well tested, we shouldn't be testing CacheControl itself.

However, any time we've written custom code around CacheControl, we need to test it, as there's infinite scope for us to mess up our own logic. Basically, we ideally want 100% test coverage of our own code.

I'm not aware of any tests that @b-m-f has written that test CacheControl endpoints - each test I've seen tests our own endpoints, and is therefore probably testing our custom logic in part and therefore valid. Could you comment more specifically on tests you don't think are necessary?

jpmartinspt commented:
```
def make_cached_session(sess,
    cache=FileStorage(".cache"),
    cache_etags=True,
    serializer=None,
    heuristic=OurDefaultHeuristic(),
    controller_class=None,
    adapter_class=None,
    cacheable_methods=None,
):
    return CacheControl(BaseSession(), *args)
```

Where we can default some of these things too other things than `None`, but still have the freedom to override whatever we need for other projects. If we have it done this way we are not adding much functionality and I would say we could probably only unit test eventual Heuristic classes that we implement.

nottrobin commented:
@jpmartinspt So there's a few things here:

We had [a discussion higher up](https://github.com/canonical-webteam/http/pull/7#discussion_r248253267) about the factory method vs class-based interface. @tbille prefers the class-based interface (which is how it used to work), and I do too as long as it can be neat. I initially didn't think it could be done neatly, but @b-m-f's use of the `CacheControlAdapter` (probably with @tbille's help) is actually nicely neat, makes is much clearer what's actually happening, and makes it easier for us to customise the `Session` as we need.

A factory to create a class is never as straightforward as creating the class directly - it obfuscates logic and the returned type from a factory function isn't inherently obvious.

Still, I believe the spirit of your point is that you want to provide the ability for the client to add custom settings to `CacheControl`. That could still be achieved in a class-based interface with something like:

``` python3
class CachedSession(BaseSession, Session):
    def __init__(
        self,
        redis_connection_pool=None,
        fallback_cache_duration=5,
        file_cache_directory=".webcache",
        cache_control_options={},
        *args,
        **kwargs
    ):
        super(BaseSession, self).__init__(*args, **kwargs)

        cc_options={
            "heuristic": ExpiresAfterIfNoCacheControl(
                seconds=fallback_cache_duration
            ),
            "cache": FileCache(file_cache_directory)
        }

        if redis_connection_pool:
            default_cache_control_options["cache"] = RedisCache(
                redis.Redis(connection_pool=redis_connection_pool)
            )

        cc_options.update(cache_control_options)

        adapter = CacheControlAdapter(**cc_options)
```

However, I'd still like to suggest that we don't do this (and maybe that we should remove the existing `*args, **kwargs` from the above constructor). I'll try to explain why.

The point of this module, it's reason to exist, is to encapsulate a set of standard settings for our projects. For the most part we want to decide those standard settings once for all projects, and discourage deviation from the standard. Therefore, I think we should deliberately restrict the number of ways in which this module can be configured - i.e. only add a configuration option when we know it's needed for solving our own problems. If we don't do this then our projects will inevitably start to diverge as they implement more and more customisations on top of our module, rendering the module kinda pointless. This module should be [strongly opinionated](https://medium.com/@stueccles/the-rise-of-opinionated-software-ca1ba0140d5b). (This is also related to the topic of [Convention of Configuration](https://en.wikipedia.org/wiki/Convention_over_configuration)).

I also believe allowing the client to pass through configuration options to one of our dependencies is a bad way to write software. Every module should be confident about its interface. We should be able to thoroughly test the functions that our module claims to provide (its public interface) and say with confidence that those functions work. If we start to allow the parts that we use internally to be customised, it becomes incredibly hard to do that - we can't possibly test that our module still works with every configuration option of both `requests.Session` and `CacheControlAdapter` - our module becomes unreliable. This is [the law of leaky abstractions](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/).

Therefore, for us to provide a solid module, we should not allow the client to configure our dependencies. Instead we should deliberately offer specific functions, and then decide ourselves, internally, how to achieve that using our dependencies.

So the client shouldn't know or care that we're using `requests.Session` or `CacheControlAdapter`, and we should have the flexibility to swap any of those pieces for other modules. All the client should care about is that `canonicalwebteam.http` provides a `CachedSession`, which can connect to a `Redis` backend and has a `fallback_cache_duration` option, and we should be able to promise that those functions will continue to work.

nottrobin commented on canonicalwebteam/http/__init__.py:
```diff
@@ -35,7 +38,7 @@
 
 
-class TimeoutHTTPAdapter(requests.adapters.HTTPAdapter):
+class TimeoutHTTPAdapter(adapters.HTTPAdapter):
```
Have we tested that the timeout still works? It needs to

b-m-f replied:
Nope. I have noted it down. But wanted to wait until our meeting to discuss the best approach.
My idea was to create a new adapter that mixes the **CacheControl** and **TimeOut** adapters.

tbille commented on canonicalwebteam/http/heuristics.py:
```diff
@@ -0,0 +1,62 @@
+        self.delta = timedelta(**kw)
+
+    def update_headers(self, response):
+        def cc_value_in_response_headers(type):
```
Be careful with naming variable `type` it's a built in python function:

https://stackoverflow.com/questions/10568087/is-it-safe-to-use-the-python-word-type-in-my-code
https://docs.python.org/3/library/functions.html#type

tbille commented on canonicalwebteam/http/heuristics.py:
```diff
@@ -0,0 +1,62 @@
+        }
+
+    def warning(self, response):
+        tmpl = "110 - Automatically cached for %s. Response might be stale"
```
Please use entire name variables, it makes it harder to read.

tbille commented on canonicalwebteam/http/heuristics.py:
```diff
@@ -0,0 +1,62 @@
+            "cache-control": "public",
+        }
+
+    def warning(self, response):
```
you are not using `response`, could be removed

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,246 @@
+        time.sleep(0.7)
+        response_3 = session.get("https://now.httpbin.org")
+
+        self.assertEqual(response_1.text, response_2.text)
```
using time.sleep can be tricky I would be careful with that:

```
======================================================================
FAIL: test_custom_heuristic (test_cache.TestCachedSession)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/tbmb/.pyenv/versions/3.6.5/envs/http/lib/python3.6/site-packages/httpretty-0.9.6-py3.6.egg/httpretty/core.py", line 1702, in wrapper
    return test(*args, **kw)
  File "/home/tbmb/Documents/projects/http/tests/test_cache.py", line 71, in test_custom_heuristic
    self.assertEqual(response_1.text, response_2.text)
AssertionError: '{"epoch": 1548239344.9905274}' != '{"epoch": 1548239346.5064354}'
- {"epoch": 1548239344.9905274}
?                     ^^^^ --
+ {"epoch": 1548239346.5064354}
?                    +++++ ^


----------------------------------------------------------------------
```

b-m-f replied:
As discussed I have changed this to use [freezegun](https://github.com/spulec/freezegun). It is much more reliable and quicker :)

b-m-f commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +103,44 @@ class UncachedSession(BaseSession, requests.Session):
+        cache = None
+
+        if redis_connection_pool:
+            r = redis.Redis(connection_pool=redis_connection_pool)
```
Remove redis dependency

b-m-f commented on canonicalwebteam/http/heuristics.py:
```diff
@@ -0,0 +1,62 @@
+                response.headers, type
+            )
+
+        if cc_value_in_response_headers(
```
If we get any Cache instructions, drop adding headers.

tbille commented on tests/test_cache.py:
```diff
@@ -0,0 +1,246 @@
+from canonicalwebteam.http import CachedSession
+from datetime import timedelta
+from mockredis import mock_redis_client
+from unittest.mock import patch
```
```
import json
import os
import pickle
import shutil
import struct
import time
import unittest
from datetime import timedelta
from unittest.mock import patch

import requests

import httpretty
import redis
from cachecontrol.caches.file_cache import FileCache
from cachecontrol.caches.redis_cache import RedisCache
from canonicalwebteam.http import CachedSession
from mockredis import mock_redis_client
```

nottrobin replied:
https://www.python.org/dev/peps/pep-0008/#imports

> Imports should be grouped in the following order:
> 
> 1. Standard library imports.
> 2. Related third party imports.
> 3. Local application/library specific imports.
> 
> You should put a blank line between each group of imports.

b-m-f committed:
Updates CachedSession with timeout for outgoing requests, removes dependency on redis

b-m-f committed:
Sorts imports

b-m-f committed:
Cleans code from unused vars. Drops any custom heuristics if other cache directives are present

b-m-f committed:
Improves heuristics and adds testing

b-m-f committed:
Uses freezegun to test time sepcific caching behaviour

nottrobin commented on canonicalwebteam/http/__init__.py:
```diff
@@ -100,24 +102,62 @@ class UncachedSession(BaseSession, requests.Session):
+
+    def __init__(self, heuristic, cache, timeout=None, *args, **kwargs):
+        self.timeout = timeout
+        self.cache_control_adapter = CacheControlAdapter(
```
Here you're creating a class that is extended from `CacheControlAdapter`, but then actually creating the controller as a property of the same class. That's weird.

What you should be doing here is using the parent (`CacheControlAdapter`)'s constructor to set up your extended object properly:

``` python3
class CacheAdapterWithTimeout(CacheControlAdapter):
    """
    A combination of the CacheControl and Timeout adapter
    to provide both functionalities.
    """

    def __init__(self, cache, heuristic, timeout=None, *args, **kwargs):
        self.timeout = timeout
        return super(CacheAdapterWithTimeout, self).__init__(
            cache=cache, heuristic=heuristic, *args, **kwargs
        )

    def send(self, *args, **kwargs):
        kwargs['timeout'] = self.timeout
        return super(CacheAdapterWithTimeout, self).send(*args, **kwargs)
```

(The `super(CacheAdapterWithTimeout, self).__init__()` syntax is ugly. It only needs to be so verbose to support Python 2 - once we can drop Python 2 support, this can become simply `super().__init__()`, which is nicer)

b-m-f committed:
Makes tests runnable with setup.py

b-m-f committed:
Readds missing parameter to warning function

nottrobin commented on canonicalwebteam/http/heuristics.py:
```diff
@@ -0,0 +1,77 @@
+            "cache-control": "public",
+        }
+
+    def warning(self,response):
```
```suggestion
    def warning(self, response):
```

nottrobin replied:
you really need a PEP8 checker in your editor :p

b-m-f replied:
I hope its now set up correctly :D

b-m-f committed:
Adds a test for time out, cleans up files

nottrobin commented on .vscode/settings.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Remove this please

nottrobin commented on tests/test_cache.py:
```diff
@@ -0,0 +1,272 @@
+
+    def test_timeout_adapter(self):
+        session = CachedSession(
+            timeout=1.2, file_cache_directory=file_cache_directory
```
In every other way, this now looks perfect. But unfortunately I'm still experiencing intermittent failures with this test.

I know it sucks, but could you just make this 2 seconds, and then use `delay/3` and `delay/1` for the test below? The values for me have to be this high to avoid intermittent failures.

If we still see issues down the line, we might just have to disable this test until we find a better way to do it.

b-m-f replied:
Have updated it  and I created an issue at https://github.com/canonical-webteam/http/issues/9 . 

Maybe I can label it with **Help wanted**?

tbille commented:
@b-m-f  `test` file is back

b-m-f committed:
Increases delays for timeout test to make it more consistent

b-m-f committed:
Removes test file

nottrobin reviewed (approved):
LGTM :+1: truly heroic effort. Thanks so much @b-m-f.
